### DISCUSSION TITLE
Weird "start_angle" / "end_angle" behavior for arcs

### DISCUSSION THREAD

**Discussion by git-n-pissed** [mozman/ezdxf#659] [Q&A]
I'm seeing some weird behavior regarding the `start_angle` and `end_angle` of arcs that doesn't seem to align with what is written in the documentation and possibly isn't consistent across DXF files.  All the arcs I'm dealing with are from 2D drawings which were generated by SolidWorks 2020.  All DXF files I'm dealing with contain a single layer, which is the "Front Plane" in SolidWorks.

Here is one arc I'm dealing with:
![arc_1](https://user-images.githubusercontent.com/52613263/159575663-abe992d5-54b9-41c2-9cfd-9c3d1b9ae05e.png)

And here are the properties that ezdxf gives me for that arc:
```
e = {Arc} ARC(#81)
    DEFAULT_ATTRIBS = {dict: 1} {'layer': '0'}
    DXFATTRIBS = {DXFAttributes} <ezdxf.lldxf.attributes.DXFAttributes object at 0x000001FB1E1AA198>
    DXFTYPE = {str} 'ARC'
    MIN_DXF_VERSION_FOR_EXPORT = {str} 'AC1009'
    appdata = {NoneType} None
    doc = {Drawing} <ezdxf.document.Drawing object at 0x000001FB1F5D24C8>
    dxf = {DXFNamespace} <ezdxf.entities.dxfns.DXFNamespace object at 0x000001FB1F5FF648>
        center = {Vec3: 3} (0.5, 0.25, 0.0)
        color = {int} 7
        dxfattribs = {DXFAttributes} <ezdxf.lldxf.attributes.DXFAttributes object at 0x000001FB1E1AA198>
        dxftype = {str} 'ARC'
        end_angle = {float} 90.0
        handle = {str} '81'
        layer = {str} '0'
        linetype = {str} 'Continuous'
        lineweight = {int} 25
        owner = {str} '1F'
        radius = {float} 0.25
        start_angle = {float} 270.0
    end_point = {Vec3: 3} (0.5, 0.5, 0.0)
    extension_dict = {NoneType} None
    has_extension_dict = {bool} False
    has_source_block_reference = {bool} False
    is_alive = {bool} True
    is_bound = {bool} True
    is_copy = {bool} False
    is_post_transform_required = {bool} False
    is_transparency_by_block = {bool} False
    is_transparency_by_layer = {bool} True
    is_virtual = {bool} False
    origin_of_copy = {NoneType} None
    proxy_graphic = {NoneType} None
    reactors = {NoneType} None
    rgb = {NoneType} None
    source_block_reference = {NoneType} None
    source_of_copy = {NoneType} None
    start_point = {Vec3: 3} (0.49999999999999994, 0.0, 0.0)
    transparency = {float} 0.0
    uuid = {UUID} af6db3a7-673f-42ff-b64a-f204f564ff61
    xdata = {NoneType} None
```

And the arc's extrusion:
```
e.dxf.extrusion = {Vec3: 3} (0.0, 0.0, 1.0)
    angle = {float} 0.0
    angle_deg = {float} 0.0
    is_null = {bool} False
    magnitude = {float} 1.0
    magnitude_square = {float} 1.0
    magnitude_xy = {float} 0.0
    spatial_angle = {float} 1.5707963267948966
    spatial_angle_deg = {float} 90.0
    vec2 = {Vec2: 2} (0.0, 0.0)
    x = {float} 0.0
    xy = {Vec3: 3} (0.0, 0.0, 0.0)
    xyz = {tuple: 3} (0.0, 0.0, 1.0)
    y = {float} 0.0
    z = {float} 1.0
```

Note that the `start_angle` is 270 and the `end_angle` is 90.  The documentation doesn't specify where the 0/360 position of the arc is, but based on the actual drawing and what ezdxf is showing it appears that in this case the 0/360 is at the 3 o'clock position and not the 12 o'clock position as one would assume.  That means these values are 90 degrees off.  I thought this is "just how it is" and I had been compensating with this code:

```
start_angle = (e.dxf.start_angle + 90.0) % 360.0
end_angle = (e.dxf.end_angle + 90.0) % 360.0
```

But then I started seeing inconsistencies between this arc and arcs in other DXF files.  Here is one such example:
![arc_2](https://user-images.githubusercontent.com/52613263/159578205-2892a228-22b9-4115-919e-839ebde26645.png)

And here are the properties that ezdxf gives me for that arc:
```
e = {Arc} ARC(#105)
    DEFAULT_ATTRIBS = {dict: 1} {'layer': '0'}
    DXFATTRIBS = {DXFAttributes} <ezdxf.lldxf.attributes.DXFAttributes object at 0x000001FB1E1AA198>
    DXFTYPE = {str} 'ARC'
    MIN_DXF_VERSION_FOR_EXPORT = {str} 'AC1009'
    appdata = {NoneType} None
    doc = {Drawing} <ezdxf.document.Drawing object at 0x000001FB1DD63188>
    dxf = {DXFNamespace} <ezdxf.entities.dxfns.DXFNamespace object at 0x000001FB23FB1B48>
        center = {Vec3: 3} (0.8877874104, 2.4064960627, 0.0)
        color = {int} 7
        dxfattribs = {DXFAttributes} <ezdxf.lldxf.attributes.DXFAttributes object at 0x000001FB1E1AA198>
        dxftype = {str} 'ARC'
        end_angle = {float} 269.9488073077
        handle = {str} '105'
        layer = {str} '0'
        linetype = {str} 'Continuous'
        lineweight = {int} 25
        owner = {str} '1F'
        radius = {float} 0.0314960629921264
        start_angle = {float} 180.0
    end_point = {Vec3: 3} (0.8877592692688279, 2.375000012279658, 0.0)
    extension_dict = {NoneType} None
    has_extension_dict = {bool} False
    has_source_block_reference = {bool} False
    is_alive = {bool} True
    is_bound = {bool} True
    is_copy = {bool} False
    is_post_transform_required = {bool} False
    is_transparency_by_block = {bool} False
    is_transparency_by_layer = {bool} True
    is_virtual = {bool} False
    origin_of_copy = {NoneType} None
    proxy_graphic = {NoneType} None
    reactors = {NoneType} None
    rgb = {NoneType} None
    source_block_reference = {NoneType} None
    source_of_copy = {NoneType} None
    start_point = {Vec3: 3} (0.8562913474078736, 2.4064960627, 0.0)
    transparency = {float} 0.0
    uuid = {UUID} 098f1cd6-0777-4ef0-89a2-e8c8d596344b
    xdata = {NoneType} None
```

And the arc's extrusion (identical to the first arc's extrusion):
```
e.dxf.extrusion = {Vec3: 3} (0.0, 0.0, 1.0)
    angle = {float} 0.0
    angle_deg = {float} 0.0
    is_null = {bool} False
    magnitude = {float} 1.0
    magnitude_square = {float} 1.0
    magnitude_xy = {float} 0.0
    spatial_angle = {float} 1.5707963267948966
    spatial_angle_deg = {float} 90.0
    vec2 = {Vec2: 2} (0.0, 0.0)
    x = {float} 0.0
    xy = {Vec3: 3} (0.0, 0.0, 0.0)
    xyz = {tuple: 3} (0.0, 0.0, 1.0)
    y = {float} 0.0
    z = {float} 1.0
```

Note that the `start_angle` is 180 and the `end_angle` is 270, but based on what I saw from the previous arc this is not what I was expecting.

The documentation which claims arcs are drawn from `start_angle` to `end_angle` in a counter-clockwise direction, but the only way that can be true for both of the example arcs I gave is for the 0/360 position to be at 3 o'clock and degree marks increment counter-clockwise.

What am I missing here?

**UPDATE:**
I've done some more investigating with additional files and it does seem that ezdxf consistently has the 0/360 position at 3 o'clock and the degree markers always increment counter-clockwise.  Not sure if there is a reason for this, but it is extremely unintuitive.

For those of you who, like me, expect 0/360 to be at the 12 o'clock position and have degree markers increment clockwise, here is some code that will accomplish that:

```
start_angle = (360 - ((e.dxf.end_angle + 270) % 360)) % 360
end_angle = (360 - ((e.dxf.start_angle + 270) % 360)) % 360
```

Here is a before and after example of that code:

![degree_markers](https://user-images.githubusercontent.com/52613263/159620538-4c8c1632-01c0-4c25-b556-a2c4a977a353.png)

**Comment by mozman** [comment]
That is the "usual" angle orientation, where 0 degree is "east" (3 o'clock) or the x-axis in the coordinate system. This is defined by the DXF reference or more precise: how AutoCAD defines and expects angle values.

The imported DXF file by SolidWorks is already an interpretation of the DXF content and not a good starting point for understanding the DXF format.

**Reply by git-n-pissed** [reply]
Yeah, I guess I do recall 0 degrees being at 3 o'clock with some Java work I did years ago.  I don't recall if the degree increments were counter-clockwise though.  That is one thing that threw me off with ezdxf.  I.e. having the arc drawn from `start_angle` to `end_angle` in the counter-clockwise direction isn't the same thing as having the degree markers incremented counter-clockwise.  Not sure if counter-clockwise degree increments are "standard" or not, but IMO it is confusing.  Might be helpful to others to clarify these points in the documentation.

I have noticed that the points in a SolidWorks drawings and the points in a DXF generated by SolidWorks are sometimes off when you get into the higher decimal places (six or more).  SolidWorks has some DXF export options that may improve that, but I haven't played with them.

Anyway, even with these issues, ezdxf has made my project much easier.  Thanks a lot for your work on it!
